CONTRIBUTING.md

```diff
@@ -268,20 +268,27 @@ also run successfully. Before you open a new pull request for your change,
 you'll want to run the test suite locally.
 
 The easiest way to run the test suite is to use
-[**tox**](https://tox.readthedocs.io/en/latest/#). You can install tox
-with pip: `pip install -U tox`. Tox provides several advantages, but the
-biggest one is that it builds an isolated virtualenv for running tests. This
-means it does not pollute your system python when running. Additionally, the
-environment that tox sets up matches the CI environment more closely and it
-runs the tests in parallel (resulting in much faster execution). To run tests
-on all installed supported python versions and lint/style checks you can simply
-run `tox`. Or if you just want to run the tests once run for a specific python
-version: `tox -epy37` (or replace py37 with the python version you want to use,
-py35 or py36).
+[**tox**](https://tox.readthedocs.io/en/latest/#). You can install tox with pip:
+`pip install -U tox`. Tox provides several advantages, but the biggest one is
+that it builds an isolated virtualenv for running tests. This means it does not
+pollute your system Python when running. Additionally, the environment that tox
+sets up matches the CI environment more closely and it runs the tests in
+parallel (resulting in much faster execution). In addition to the requirements
+installed by `tox` in the virtualenv, you'll need to install the LaTeX packages `xpic`,
+`qcircuit`, and `standalone`. These are typically provided in Linux distributions
+by packages of the same names prefixed by `texlive-`. For MacOS a good option is
+[MacTeX](https://www.tug.org/mactex/).
```

Comment:
Is a latex distribution actually required for running the tests successfully? (I've never actually tried as every system I've run the tests on had texlive installed well before Qiskit was cloned) If so I actually treat this as a bug in the test suite as latex is an optional dependency and we shouldn't be requiring people to install it to pass tests (it should just skip if latex isn't present).

That being said we should still mention latex as an optional thing here so it's clear to run the latex tests people need to have it installed. Similarly there are other binary programs which are optional dependencies for things, the big example I can think of is graphviz which trips up a lot of people. There is also poppler which is needed to convert the pdf to an image file for the latex drawer.

Reply:
Tests are skipped if there's no `pdflatex` binary available, but the point is that it's possible to have a `pdflatex` binary without the TeX packages `xypic` (note there's typo in this diff) and `qcircuit` if you chose a very minimal LaTeX installation.  We don't test for the presence of those directly; the LaTeX output will just fail if they're missing.  That was the discussion from the meeting.

It's possible to add tests for them using a further dependency on the `kpsewhich` binary which is _only_ used to search the LaTeX import path for those files, but that carries a risk that all the packages actually were available but `kpsewhich` wasn't.  `kpsewhich` is only a natural part of TeX Live (the most common base on Unix-likes) not MiKTeX (most common distribution on Windows), but MiKTeX does include emulation of kpathsea and its components like `kpsewhich` which has sometimes gone awry.  I don't know about other LaTeX distributions.  My argument from the meeting was that I think relying on `kpsewhich` being available is an unnecessary extra point of failure for users that can actively prevent the otherwise-usable LaTeX components from being available, and it might be more suitable just to let the error message say "have you made sure `xypic`, `standalone` and `qcircuit` are available?".

That wouldn't let us auto-skip tests, but my personal vote is prefer "no innocent user gets locked out" over "no developer sees an unnecessary but actionable error".  _That said_, the way the lazy testers are designed is for subclassing.  I just made two concrete subclasses `LazyImportTester` and `LazySubprocessTester` for the use-cases we needed.  If we decided that we want to use `kpsewhich` for further testing iff it's available, it's easy to write a specific subclass of `_LazyTester` just to handle that very special case.

Reply:
I didn't realize that the test is skipped if `pdflatex` is not found. So this has to be reworded somehow. I agree we should prioritize not locking users out. Regarding testing, the hassle that I had running the test suite would have been solved easily simply by having the latex packages documented somehow here. (I don't know if this would have solved @javabster 's issue, since I think she was not running tests when the problem arose.) Having a precise error message pop out so you don't have to grep `CONTRIBUTING.md` would be nice, but not a huge time saver.

Reply:
I wouldn't be surprised if `pdflatex` differs between the TeX distributions, eg in command line switches. Trying to determine if the packages are present using `pdflatex` is a bit more fiddly than using something like `kpswhich`. But it does have the advantage of not relying on the presence of another executable. I think you wouldn't know which works better in the wild until you try.

After researching TeX distributions and their history, I suspect that a user having `pdflatex` on their path, but not `kpsewhich` would happen rarely if ever. I'm unable to find a distribution in use that does not include `kpsewhich`. We can skip the test if `kpsewhich` is missing in which case there would be no failure. We could also call `findtexmf` from MiKTeX.  In this case you might want to test with `pdflatex` as a backup--- but why not just do that in the first place?

Some relevant culture. My best understanding is that the tool `kpsewhich` appeared at the same time as `kpathsea` the libarary that it depends on.  I found source files showing that `kpsewhich` appeared no later than 1994 in TeTeX. The author of MiKTeX said that path searching was based on `kpathsea`, but of course was not compatible. I don't know when `kpsewhich` first appeared in MiKTeX. Although MiKTeX first appeared in the early 90's, it seems all traces of source code, changelogs, commit messages, version numbers, etc. from before 2016 have disappeared from the internet. My guess is that `kpsewhich` is essential to MiKTeX broadly construed, as it is called by several tools included in the source distribution. I installed MiKTeX on linux from source. The installation and instructions are broken and out of date. The documentation in general is thin. `pdflatex` is missing, but `kpsewhich` does work. I am able to check that it's behavior differs at least in some small ways from the one in texlive. But for the case at hand, it is not important.

Reply:
I made the discussion more accurate in 827a15394d7c2fc1ea0cd32b10cd567f97733102